---
title: "STR graphs"
output: html_document
---

```{r, include=FALSE}
str_dat <- read.csv(file="BracketNotation_A4-C-WA29594_S25_L001_R1_001.fastq.STRaitRazor_output.csv", header=TRUE, sep=",", stringsAsFactors = FALSE)

library(dplyr)
library(ggplot2)

tbl_df(str_dat)

str_dat2 <- slice(str_dat, 1:44)

str_dat2$Locus <- as.character(str_dat2$Locus)

suppressWarnings(str_dat2$Coverage.of.majority.peaks <- as.numeric(as.character(str_dat2$Coverage.of.majority.peaks)))

str_dat2[is.na(str_dat2)] <- 0 

str_dat2 <- str_dat2  %>%  group_by(Locus)  %>%  mutate(sum_coverage=sum(Coverage.of.majority.peaks))

```

```{r, fig.height=7, fig.width=7, echo=FALSE}
het_dat <- str_dat2 %>% group_by(Locus) %>% 
              mutate(ratio = Coverage.of.majority.peaks/sum_coverage) %>% 
              top_n(1, Coverage.of.majority.peaks)
## group_by locus
## id max allele
## summarise(ratio = max(coverage)/sum(coverage), max_allele = )
ggplot(data=het_dat, aes(x=Locus, y=ratio, color = Locus)) + 
  geom_point(position = "identity", size=3) + 
  guides(colour=FALSE) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  geom_hline(aes(yintercept = 0.65), linetype = 2, color = "grey80") +
  ylab("Ratio") + 
  ylim(0,1) +
  ggtitle("Heterozygote Balance")
```

##### This graph shows the proportion of reads of of each allele for each individual locus.


```{r, echo=FALSE, fig.height = 5}
ggplot(data=str_dat2, aes(x=Locus, y=sum_coverage, colour=Locus)) + 
  geom_point(position = "identity", size=3) + 
  guides(colour=FALSE) +
  ylab("Total Coverage") +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  xlab("Locus") +
  ggtitle("Interlocus Balance") +
  theme(aspect.ratio = .125) +
  geom_hline(aes(yintercept=median(str_dat2$sum_coverage)), linetype = 2)#, color = "grey80")
```

##### This graph shows the comparison of the total coverage across all the loci with the horizontal line representing the median coverage value.

