---
title: "STRaitrazor Analysis"
output: html_document
---
### Data source
Summary analysis of data from STRaitrazor, a software that detects STR alleles from raw sequencing data. The DNA was sequenced both in the forward and reverse directions, and since the ends were paired, each was assigned a different read. The raw output from STRaightrazor contained information on the locus, the allele number, the number of bases, the coverage value, and the actual STR sequence itself.

#### Locus Counts
For each locus, the alleles that had coverage hits were graphed for each read, and compared side by side for comparison. 


```{r, message=FALSE, echo = FALSE, warning=FALSE}
#This will call the file that contains the functions for uploading and processing the data
source("process_straitrazor_output.R")
``` 


```{r, warning=FALSE, error=FALSE, echo=FALSE, message=FALSE}
#This will create two data frames, one for the . sequences files and one for the .allelcalls files
seq_df <- process_sequences_directory("H12-H-ZT80925_S96_L001_R1_001.fastq.STRaitRazor/PE_MiSeq_ALL/")
allelecalls_df <- process_allelecalls_directory("H12-H-ZT80925_S96_L001_R1_001.fastq.STRaitRazor/PE_MiSeq_ALL/")
```


```{r, echo=FALSE, message=FALSE}
library(stringr)
allele_lables <- seq_df$Allele %>% 
                    str_replace("X", "100") %>% 
                    str_replace("Y", "1010") %>% 
                    as.numeric() %>% 
                    unique() %>% 
                    sort() %>% 
                    as.character() %>% 
                    str_replace("100", "X") %>% 
                    str_replace("1010", "Y")
seq_df %<>% mutate(Allele_factored = factor(Allele, levels = allele_lables))
```


```{r fig.height = 10, echo=FALSE}
seq_df %>% filter(Allele_Calls != 1) %>% 
    ggplot(aes(Allele_factored, Allele_Calls, fill=Read)) +
    geom_bar(position="dodge", stat="identity") +
    facet_wrap(~Locus, scale = "free_x",ncol = 3) + 
    ylab("Coverage Count") +
    xlab("Allele") +
    ggtitle("Coverage Value by Locus") +
    theme_bw()
```

```{r, include=FALSE}
str_dat <- read.csv(file="BracketNotation_A4-C-WA29594_S25_L001_R1_001.fastq.STRaitRazor_output.csv", header=TRUE, sep=",", stringsAsFactors = FALSE)

library(dplyr)
library(ggplot2)

tbl_df(str_dat)

str_dat2 <- slice(str_dat, 1:44)

str_dat2$Locus <- as.character(str_dat2$Locus)

suppressWarnings(str_dat2$Coverage.of.majority.peaks <- as.numeric(as.character(str_dat2$Coverage.of.majority.peaks)))

str_dat2[is.na(str_dat2)] <- 0 

str_dat2 <- str_dat2  %>%  group_by(Locus)  %>%  mutate(sum_coverage=sum(Coverage.of.majority.peaks))

```


#### Read proportion by Locus
This graph shows the proportion of reads of of each allele for each individual locus.

```{r, fig.height=7, fig.width=7, echo=FALSE}
het_dat <- str_dat2 %>% group_by(Locus) %>% 
              mutate(ratio = Coverage.of.majority.peaks/sum_coverage) %>% 
              top_n(1, Coverage.of.majority.peaks)
ggplot(data=het_dat, aes(x=Locus, y=ratio, color = Locus)) + 
  geom_point(position = "identity", size=3) + 
  guides(colour=FALSE) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  geom_hline(aes(yintercept = 0.65), linetype = 2, color = "grey80") +
  ylab("Ratio") + 
  ylim(0,1) +
  ggtitle("Heterozygote Balance")
```



#### Interlocus Balance
This graph shows the total number of reads as a comparison between loci.

```{r, echo=FALSE, fig.height = 7}
ggplot(data=str_dat2, aes(x=Locus, y=sum_coverage, colour=Locus)) + 
  geom_point(position = "identity", size=3) + 
  guides(colour=FALSE) +
  ylab("Total Coverage") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  xlab("Locus") +
  ggtitle("Interlocus Balance") +
  geom_hline(aes(yintercept=median(str_dat2$sum_coverage)), linetype = 2, color = "grey80") 
```


######These tables show the information graphs as well as additional data recorded.

## Table 1


```{r, echo=FALSE}
het_dat2 <- het_dat[,-c(3:13)]
het_dat2 <- rename(het_dat2, "Total Coverage" = sum_coverage)
het_dat2 <- rename(het_dat2, "Heterozygote Balance" = ratio)
het_dat2[,4] <- round(het_dat2[,4], digits = 4)
library(DT)
datatable(het_dat2)

```

## Table 2

```{r, echo=FALSE}
str_dat2 <- str_dat2[,-c(5,6,8,14)]
str_dat2 <- rename(str_dat2, "Forward Bracket Notation" = Foward.Bracket.Notation)
str_dat2 <- rename(str_dat2, "Reverse Bracket Notation" = Reverse.Bracket.Notation)
str_dat2 <- rename(str_dat2, "Reverse Complement" = Reverse.complement.)
str_dat2 <- rename(str_dat2, "Read Bias" = Read.bias)
str_dat2 <- rename(str_dat2, "Strand Bias" = Strand.bias)
str_dat2 <- rename(str_dat2, "Stutter Ratio" = Stutter.ratio)
str_dat2 <- rename(str_dat2, "Stutter Coverage" = Stutter.coverage)
str_dat2 <- rename(str_dat2, "Percentage of Non Majority Peaks" = Percentage.of.non.majority.peaks)
datatable(str_dat2)
```

